File: Sesion 8/Ejercicio/Ejercicio1 S8/EJERCICIO 1 S8.cpp
#include <iostream>
using namespace std;

const double PI = 3.1416; 

// Clase abstracta Padre Figura
class Figura {
public:
    virtual void dibujar() = 0;
    virtual double calcularArea() = 0;
};


// Clase derivada Círculo
class Círculo : public Figura {
private:
    double radio;

public:
    Círculo(double r) : radio(r) {}

    void dibujar() {
        cout << "Dibujando un círculo" << endl;
    }

    double calcularArea() {
        return PI * radio * radio;
    }
};


// Clase derivada Rectángulo
class Rectángulo : public Figura {
private:
    double base, altura;

public:
    Rectángulo(double b, double h) : base(b), altura(h) {}

    void dibujar() {
        cout << "Dibujando un rectángulo" << endl;
    }

    double calcularArea() {
        return base * altura;
    }
};


// Clase derivada Triangulo
class Triangulo : public Figura {
private:
    double base, altura;

public:
    Triangulo(double b, double h) : base(b), altura(h) {}

    void dibujar() {
        cout << "Dibujando un triangulo" << endl;
    }

    double calcularArea() {
        return 0.5 * base * altura;
    }
};


int main() {
    Rectángulo rectangulo(4, 6);
    rectangulo.dibujar();
    cout << "El area del rectángulo: " << rectangulo.calcularArea() << endl;

    Círculo circulo(5);
    circulo.dibujar();
    cout << "El area del círculo: " << circulo.calcularArea() << endl;

    Triangulo triangulo(3, 8);
    triangulo.dibujar();
    cout << "El area del triangulo: " << triangulo.calcularArea() << endl;

    return 0;
}

File: Sesion 8/Ejercicio/Ejercicio1 S8/EJERCICIO 1 S8.java
import java.lang.Math;

public abstract class Figura {
    public abstract void dibujar();
    public abstract double calcularArea();
}

class Círculo extends Figura {
    private double radio;

    public Círculo(double r) {
        radio = r;
    }

    public void dibujar() {
        System.out.println("Dibujando un círculo");
    }

    public double calcularArea() {
        return Math.PI * radio * radio;
    }
}

class Rectángulo extends Figura {
    private double base, altura;

    public Rectángulo(double b, double h) {
        base = b;
        altura = h;
    }

    public void dibujar() {
        System.out.println("Dibujando un rectángulo");
    }

    public double calcularArea() {
        return base * altura;
    }
}

class Triangulo extends Figura {
    private double base, altura;

    public Triangulo(double b, double h) {
        base = b;
        altura = h;
    }

    public void dibujar() {
        System.out.println("Dibujando un triangulo");
    }

    public double calcularArea() {
        return 0.5 * base * altura;
    }
}

public class Main {
    public static void main(String[] args) {
        Rectángulo rectangulo = new Rectángulo(4, 6);
        rectangulo.dibujar();
        System.out.println("El area del rectángulo: " + rectangulo.calcularArea());

        Círculo circulo = new Círculo(5);
        circulo.dibujar();
        System.out.println("El area del círculo: " + circulo.calcularArea());

        Triangulo triangulo = new Triangulo(3, 8);
        triangulo.dibujar();
        System.out.println("El area del triangulo: " + triangulo.calcularArea());
    }
}

